Looks like test-qnn-buck is failing : 
buck-out/v2/gen/root/904931f735703749/backends/qualcomm/runtime/__runtime_static__/buck-headers/executorch/backends/qualcomm/runtime/backends/QnnImplementation.h:13:10: fatal error: 'pal/DynamicLoading.h' file not found
#include <pal/DynamicLoading.h>

--- a/backends/qualcomm/CMakeLists.txt
+++ b/backends/qualcomm/CMakeLists.txt
@@ -119,6 +119,7 @@ endif()
 include_directories(
   BEFORE ${_common_include_directories} ${QNN_SDK_ROOT}/include/QNN
   ${EXECUTORCH_SOURCE_DIR}/runtime/core/portable_type/c10
+  ${QNN_EXECUTORCH_ROOT_DIR}/runtime/pal/include
 )
 
 set(_qnn_schema__srcs backends/qualcomm/serialization/qc_compiler_spec.fbs)
@@ -175,6 +176,7 @@ add_library(qnn_logger STATIC)
 add_library(qnn_manager STATIC)
 add_library(qnn_mem_manager STATIC)
 add_library(qnn_op_package_manager STATIC)
+add_library(qnn_pal STATIC)
 add_library(qnn_profiler STATIC)
 add_library(qnn_schema INTERFACE ${_qnn_schema__outputs})
 add_library(qnn_sys_function_interface INTERFACE)
@@ -188,11 +190,12 @@ add_library(wrappers STATIC)
 target_link_libraries(wrappers PRIVATE qnn_executorch_logging)
 target_link_libraries(
   qnn_implementation PRIVATE qnn_function_interface qnn_executorch_logging
-                             ${CMAKE_DL_LIBS}
+                             qnn_pal ${CMAKE_DL_LIBS}
 )
 target_link_libraries(
-  qnn_sys_implementation PRIVATE qnn_sys_function_interface
-                                 qnn_executorch_logging ${CMAKE_DL_LIBS}
+  qnn_sys_implementation
+  PRIVATE qnn_sys_function_interface qnn_executorch_logging qnn_pal
+          ${CMAKE_DL_LIBS}
 )
 target_link_libraries(qnn_executorch_logging PRIVATE qnn_schema)
 target_link_libraries(qnn_profiler PRIVATE qnn_executorch_logging)
@@ -243,6 +246,7 @@ target_link_libraries(
           qnn_context
           qnn_graph
           qnn_mem_manager
+          qnn_pal
 )
 
 target_link_libraries(
@@ -297,7 +301,7 @@ else()
   )
 endif()
 target_link_libraries(
-  shared_buffer PRIVATE qnn_executorch_logging ${CMAKE_DL_LIBS}
+  shared_buffer PRIVATE qnn_executorch_logging qnn_pal ${CMAKE_DL_LIBS}
 )
 #
 # add linker option

--- a/backends/qualcomm/runtime/CMakeLists.txt
+++ b/backends/qualcomm/runtime/CMakeLists.txt
@@ -47,3 +47,16 @@ target_sources(
   shared_buffer PRIVATE ${CMAKE_CURRENT_LIST_DIR}/SharedBuffer.h
                         ${CMAKE_CURRENT_LIST_DIR}/SharedBuffer.cpp
 )
+
+# qnn_pal
+if(WIN32)
+  target_sources(
+    qnn_pal PRIVATE ${CMAKE_CURRENT_LIST_DIR}/pal/src/windows/DynamicLoading.cpp
+                    ${CMAKE_CURRENT_LIST_DIR}/pal/src/windows/Path.cpp
+  )
+else()
+  target_sources(
+    qnn_pal PRIVATE ${CMAKE_CURRENT_LIST_DIR}/pal/src/linux/DynamicLoading.cpp
+                    ${CMAKE_CURRENT_LIST_DIR}/pal/src/linux/Path.cpp
+  )
+endif()

--- a/backends/qualcomm/runtime/SharedBuffer.cpp
+++ b/backends/qualcomm/runtime/SharedBuffer.cpp
@@ -5,9 +5,10 @@
  * This source code is licensed under the BSD-style license found in the
  * LICENSE file in the root directory of this source tree.
  */
-#include <dlfcn.h>
 #include <executorch/backends/qualcomm/runtime/Logging.h>
 #include <executorch/backends/qualcomm/runtime/SharedBuffer.h>
+
+#include <pal/DynamicLoading.h>
 
 // Refer to the QNN HTP Shared Buffer Tutorial
 // in Qualcomm® AI Engine Direct document
@@ -81,10 +82,10 @@ SharedBuffer& SharedBuffer::GetSharedBufferManager() {
   std::lock_guard<std::mutex> lk(init_mutex_);
   static SharedBuffer shared_buffer_manager;
   if (!shared_buffer_manager.GetInitialize()) {
-#if defined(__aarch64__)
+#if defined(__ANDROID__)
     Error status = shared_buffer_manager.Load();
 #else
-    // For x86_64 platform
+    // libcdsprpc.so is an Android vendor library; skip on all other platforms.
     Error status = Error::Ok;
 #endif
     if (status == Error::Ok) {
@@ -95,7 +96,7 @@ SharedBuffer& SharedBuffer::GetSharedBufferManager() {
 }
 
 SharedBuffer::~SharedBuffer() {
-#if defined(__aarch64__)
+#if defined(__ANDROID__)
   if (initialize_) {
     SharedBuffer::GetSharedBufferManager().UnLoad();
   }
@@ -163,23 +164,27 @@ bool SharedBuffer::IsAllocated(void* buf) {
 Error SharedBuffer::Load() {
   // On Android, 32-bit and 64-bit libcdsprpc.so can be found at /vendor/lib/
   // and /vendor/lib64/ respectively.
-  lib_cdsp_rpc_ = dlopen("libcdsprpc.so", RTLD_NOW | RTLD_LOCAL);
+  lib_cdsp_rpc_ = pal::dynamic_loading::DlOpen(
+      "libcdsprpc.so",
+      pal::dynamic_loading::DL_NOW | pal::dynamic_loading::DL_LOCAL);
   if (lib_cdsp_rpc_ == nullptr) {
     QNN_EXECUTORCH_LOG_ERROR(
-        "Unable to load shared buffer. dlerror(): %s", dlerror());
+        "Unable to load shared buffer. dlerror(): %s",
+        pal::dynamic_loading::DlError());
     return Error::Internal;
   }
   rpc_mem_alloc_ = reinterpret_cast<RpcMemAllocFn_t>( // NOLINT
-      dlsym(lib_cdsp_rpc_, "rpcmem_alloc"));
+      pal::dynamic_loading::DlSym(lib_cdsp_rpc_, "rpcmem_alloc"));
   rpc_mem_free_ = reinterpret_cast<RpcMemFreeFn_t>( // NOLINT
-      dlsym(lib_cdsp_rpc_, "rpcmem_free"));
+      pal::dynamic_loading::DlSym(lib_cdsp_rpc_, "rpcmem_free"));
   rpc_mem_to_fd_ = reinterpret_cast<RpcMemToFdFn_t>( // NOLINT
-      dlsym(lib_cdsp_rpc_, "rpcmem_to_fd"));
+      pal::dynamic_loading::DlSym(lib_cdsp_rpc_, "rpcmem_to_fd"));
   if (nullptr == rpc_mem_alloc_ || nullptr == rpc_mem_free_ ||
       nullptr == rpc_mem_to_fd_) {
     QNN_EXECUTORCH_LOG_ERROR(
-        "Unable to access symbols in shared buffer. dlerror(): %s", dlerror());
-    dlclose(lib_cdsp_rpc_);
+        "Unable to access symbols in shared buffer. dlerror(): %s",
+        pal::dynamic_loading::DlError());
+    pal::dynamic_loading::DlClose(lib_cdsp_rpc_);
     return Error::Internal;
   }
   return Error::Ok;
@@ -199,9 +204,10 @@ void SharedBuffer::AddCusomMemTensorAddr(void* tensor_addr, void* custom_mem) {
 };
 
 Error SharedBuffer::UnLoad() {
-  if (dlclose(lib_cdsp_rpc_) != 0) {
+  if (pal::dynamic_loading::DlClose(lib_cdsp_rpc_) != 0) {
     QNN_EXECUTORCH_LOG_ERROR(
-        "Unable to close shared buffer. dlerror(): %s", dlerror());
+        "Unable to close shared buffer. dlerror(): %s",
+        pal::dynamic_loading::DlError());
     return Error::Internal;
   };
   return Error::Ok;

--- a/backends/qualcomm/runtime/backends/QnnBackendUnifiedRegistry.h
+++ b/backends/qualcomm/runtime/backends/QnnBackendUnifiedRegistry.h
@@ -15,6 +15,8 @@
 #include <executorch/backends/qualcomm/runtime/backends/QnnSysImplementation.h>
 #include <executorch/runtime/core/error.h>
 
+#include <pal/Path.h>
+
 #include <memory>
 #include <mutex>
 #include <string>
@@ -79,12 +81,16 @@ class QnnBackendUnifiedRegistry {
   // For macro, refer to executorch/backends/qualcomm/CMakeLists.txt
   static constexpr const char* htp_library_name_ = HEXAGON_LIB;
 #else
-  static constexpr const char* htp_library_name_ = "libQnnHtp.so";
+  static inline const std::string htp_library_name_ =
+      pal::path::GetLibraryName("QnnHtp");
 #endif
-  static constexpr const char* gpu_library_name_ = "libQnnGpu.so";
-  static constexpr const char* dsp_library_name_ = "libQnnDsp.so";
+  static inline const std::string gpu_library_name_ =
+      pal::path::GetLibraryName("QnnGpu");
+  static inline const std::string dsp_library_name_ =
+      pal::path::GetLibraryName("QnnDsp");
   // Lpai library name is same for both traditional build and hexagon build.
-  static constexpr const char* lpai_library_name_ = "libQnnLpai.so";
+  static inline const std::string lpai_library_name_ =
+      pal::path::GetLibraryName("QnnLpai");
 
   std::unique_ptr<const QnnSaver_Config_t*[]> GetImplementationConfig(
       const QnnExecuTorchOptions* options) {

--- a/backends/qualcomm/runtime/backends/QnnDlcManager.h
+++ b/backends/qualcomm/runtime/backends/QnnDlcManager.h
@@ -14,6 +14,8 @@
 #include <executorch/backends/qualcomm/runtime/backends/QnnBackendUnifiedRegistry.h>
 #include <executorch/backends/qualcomm/runtime/backends/ir/IrContext.h>
 
+#include <pal/Path.h>
+
 namespace executorch {
 namespace backends {
 namespace qnn {
@@ -105,7 +107,8 @@ class QnnDlcManager {
   }
 
  private:
-  static constexpr const char* library_name_ = "libQnnIr.so";
+  static inline const std::string library_name_ =
+      pal::path::GetLibraryName("QnnIr");
 
   const QnnExecuTorchContextBinary& qnn_context_blob_;
   const QnnExecuTorchOptions* options_;

--- a/backends/qualcomm/runtime/backends/QnnImplementation.cpp
+++ b/backends/qualcomm/runtime/backends/QnnImplementation.cpp
@@ -22,7 +22,7 @@ struct DlCloser {
   int operator()(void* handle) {
     if (handle == nullptr)
       return 0;
-    return dlclose(handle);
+    return pal::dynamic_loading::DlClose(handle);
   }
 };
 
@@ -59,22 +59,26 @@ const QnnInterface_t* QnnImplementation::StartBackend(
 
 #ifdef __hexagon__
   FARF(RUNTIME_HIGH, "Opening lib_path %s", lib_path.c_str());
-  std::unique_ptr<void, DlCloser> lib_handle(
-      dlopen(lib_path.c_str(), RTLD_NOW | RTLD_GLOBAL));
+  std::unique_ptr<void, DlCloser> lib_handle(pal::dynamic_loading::DlOpen(
+      lib_path.c_str(),
+      pal::dynamic_loading::DL_NOW | pal::dynamic_loading::DL_GLOBAL));
   FARF(RUNTIME_HIGH, "Done loading lib_path %s", lib_path.c_str());
 #else
   // If the library is already loaded, return the handle.
-  std::unique_ptr<void, DlCloser> lib_handle(
-      dlopen(lib_path.c_str(), RTLD_NOW | RTLD_NOLOAD | RTLD_GLOBAL));
+  std::unique_ptr<void, DlCloser> lib_handle(pal::dynamic_loading::DlOpen(
+      lib_path.c_str(),
+      pal::dynamic_loading::DL_NOW | pal::dynamic_loading::DL_NOLOAD |
+          pal::dynamic_loading::DL_GLOBAL));
   if (!lib_handle) {
-    lib_handle = std::unique_ptr<void, DlCloser>(
-        dlopen(lib_path.c_str(), RTLD_NOW | RTLD_GLOBAL));
+    lib_handle = std::unique_ptr<void, DlCloser>(pal::dynamic_loading::DlOpen(
+        lib_path.c_str(),
+        pal::dynamic_loading::DL_NOW | pal::dynamic_loading::DL_GLOBAL));
   }
   if (lib_handle == nullptr) {
     QNN_EXECUTORCH_LOG_ERROR(
         "Cannot Open QNN library %s, with error: %s",
         lib_path.c_str(),
-        dlerror());
+        pal::dynamic_loading::DlError());
     return nullptr;
   }
 #endif
@@ -87,7 +91,7 @@ const QnnInterface_t* QnnImplementation::StartBackend(
     QNN_EXECUTORCH_LOG_ERROR(
         "QnnImplementation::Load Cannot load symbol "
         "QnnInterface_getProviders : %s",
-        dlerror());
+        pal::dynamic_loading::DlError());
     return nullptr;
   }
 
@@ -131,12 +135,12 @@ Error QnnImplementation::Unload() {
     return Error::Ok;
   }
 
-  int dlclose_error = dlclose(lib_handle_);
+  int dlclose_error = pal::dynamic_loading::DlClose(lib_handle_);
   if (dlclose_error != 0) {
     QNN_EXECUTORCH_LOG_ERROR(
         "Fail to close QNN backend %s with error %s",
         lib_path_.c_str(),
-        dlerror());
+        pal::dynamic_loading::DlError());
     return Error::Internal;
   }
   lib_handle_ = nullptr;

--- a/backends/qualcomm/runtime/backends/QnnImplementation.h
+++ b/backends/qualcomm/runtime/backends/QnnImplementation.h
@@ -10,7 +10,8 @@
 #include <executorch/backends/qualcomm/runtime/Logging.h>
 #include <executorch/backends/qualcomm/runtime/backends/QnnFunctionInterface.h>
 
-#include <dlfcn.h>
+#include <pal/DynamicLoading.h>
+
 #include <string>
 namespace executorch {
 namespace backends {
@@ -18,7 +19,8 @@ namespace qnn {
 
 template <typename Fn>
 Fn loadQnnFunction(void* handle, const char* function_name) {
-  return reinterpret_cast<Fn>(dlsym(handle, function_name)); // NOLINT
+  return reinterpret_cast<Fn>( // NOLINT
+      pal::dynamic_loading::DlSym(handle, function_name));
 }
 
 class QnnImplementation {

--- a/backends/qualcomm/runtime/backends/QnnSysImplementation.cpp
+++ b/backends/qualcomm/runtime/backends/QnnSysImplementation.cpp
@@ -6,8 +6,8 @@
  * LICENSE file in the root directory of this source tree.
  */
 
-#include <dlfcn.h>
 #include <executorch/backends/qualcomm/runtime/backends/QnnSysImplementation.h>
+#include <pal/DynamicLoading.h>
 namespace executorch {
 namespace backends {
 namespace qnn {
@@ -16,24 +16,27 @@ using executorch::runtime::Error;
 
 Error QnnSystemImplementation::Load() {
   Qnn_ErrorHandle_t error = QNN_SUCCESS;
-  void* lib_handle_ = dlopen(lib_path_.c_str(), RTLD_NOW | RTLD_LOCAL);
+  lib_handle_ = pal::dynamic_loading::DlOpen(
+      lib_path_.c_str(),
+      pal::dynamic_loading::DL_NOW | pal::dynamic_loading::DL_LOCAL);
   if (lib_handle_ == nullptr) {
     QNN_EXECUTORCH_LOG_ERROR(
         "Cannot Open QNN library %s, with error: %s",
         lib_path_.c_str(),
-        dlerror());
+        pal::dynamic_loading::DlError());
     return Error::Internal;
   }
 
   auto* get_providers =
       // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
       reinterpret_cast<QnnSystemInterfaceGetProvidersFn*>(
-          dlsym(lib_handle_, "QnnSystemInterface_getProviders"));
+          pal::dynamic_loading::DlSym(
+              lib_handle_, "QnnSystemInterface_getProviders"));
   if (get_providers == nullptr) {
     QNN_EXECUTORCH_LOG_ERROR(
         "QnnSystemImplementation::Load Cannot load symbol "
         "QnnSystemInterface_getProviders : %s",
-        dlerror());
+        pal::dynamic_loading::DlError());
     return Error::Internal;
   }
 
@@ -65,10 +68,11 @@ Error QnnSystemImplementation::Unload() {
   if (lib_handle_ == nullptr)
     return Error::Ok;
 
-  int dlclose_error = dlclose(lib_handle_);
+  int dlclose_error = pal::dynamic_loading::DlClose(lib_handle_);
   if (dlclose_error != 0) {
     QNN_EXECUTORCH_LOG_WARN(
-        "Failed to close QnnSystem library with error %s", dlerror());
+        "Failed to close QnnSystem library with error %s",
+        pal::dynamic_loading::DlError());
     return Error::Internal;
   }
 

--- a/backends/qualcomm/runtime/backends/ir/target/QnnDlcManager.cpp
+++ b/backends/qualcomm/runtime/backends/ir/target/QnnDlcManager.cpp
@@ -6,10 +6,12 @@
  * LICENSE file in the root directory of this source tree.
  */
 #include <executorch/backends/qualcomm/runtime/backends/QnnDlcManager.h>
+#ifndef _WIN32
 #include <fcntl.h>
 #include <sys/mman.h>
 #include <sys/stat.h>
 #include <unistd.h>
+#endif
 #include <fstream>
 
 namespace executorch {

--- a/backends/qualcomm/runtime/pal/include/pal/DynamicLoading.h
+++ b/backends/qualcomm/runtime/pal/include/pal/DynamicLoading.h
@@ -1,0 +1,79 @@
+/*
+ * Copyright (c) Qualcomm Innovation Center, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+//---------------------------------------------------------------------------
+/// @file
+///   This file includes APIs for dynamic loading on supported platforms
+//---------------------------------------------------------------------------
+
+#pragma once
+
+namespace pal {
+namespace dynamic_loading {
+
+enum {
+  DL_NOW = 0x0001,
+  DL_LOCAL = 0x0002,
+  DL_GLOBAL = 0x0004,
+  DL_NOLOAD = 0x0008
+};
+
+//---------------------------------------------------------------------------
+/// @brief
+///   Loads the dynamic shared object
+/// @param filename
+///   If contains path separators, treat it as relative or absolute pathname
+///   or search it for the rule of dynamic linker
+/// @param flags
+///   - DL_NOW: resolve undefined symbols before return. MUST be specified.
+///   - DL_LOCAL: symbols defined in this shared object are not made available
+///     to resolve references in subsequently loaded shared objects
+///   - DL_GLOBAL: resolve symbol globally
+///   - DL_NOLOAD: only test if the library is loaded, return handle if so
+/// @return
+///   On success, a non-NULL handle for the loaded library.
+///   On error, NULL
+//---------------------------------------------------------------------------
+void* DlOpen(const char* filename, int flags);
+
+//---------------------------------------------------------------------------
+/// @brief
+///   Obtain address of a symbol in a shared object or executable
+/// @param handle
+///   A handle of a dynamic loaded shared object returned by DlOpen
+/// @param symbol
+///   A null-terminated symbol name
+/// @return
+///   On success, return the address associated with symbol
+///   On error, NULL
+//---------------------------------------------------------------------------
+void* DlSym(void* handle, const char* symbol);
+
+//---------------------------------------------------------------------------
+/// @brief
+///   Decrements the reference count on the dynamically loaded shared object
+///   referred to by handle.
+/// @param handle
+///   A handle of a dynamic loaded shared object returned by DlOpen
+/// @return
+///   On success, 0; on error, a nonzero value
+//---------------------------------------------------------------------------
+int DlClose(void* handle);
+
+//---------------------------------------------------------------------------
+/// @brief
+///   Obtain error diagnostic for functions in the dl-family APIs.
+/// @return
+///   Returns a human-readable, null-terminated string describing the most
+///   recent error that occurred from a call to one of the functions in the
+///   dl-family APIs.
+//---------------------------------------------------------------------------
+const char* DlError();
+
+} // namespace dynamic_loading
+} // namespace pal

--- a/backends/qualcomm/runtime/pal/include/pal/Path.h
+++ b/backends/qualcomm/runtime/pal/include/pal/Path.h
@@ -1,0 +1,34 @@
+/*
+ * Copyright (c) Qualcomm Innovation Center, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+//---------------------------------------------------------------------------
+/// @file
+///   This file includes APIs for path related operations on supported platforms
+//---------------------------------------------------------------------------
+
+#pragma once
+
+#include <string>
+
+namespace pal {
+namespace path {
+
+//---------------------------------------------------------------------------
+/// @brief
+///   Returns the platform-specific shared library filename for the given
+///   base name. For example, "QnnHtp" becomes "libQnnHtp.so" on Linux
+///   or "QnnHtp.dll" on Windows.
+/// @param baseName
+///   The base name of the library without prefix or extension
+/// @return
+///   The platform-specific library filename
+//---------------------------------------------------------------------------
+std::string GetLibraryName(const std::string& baseName);
+
+} // namespace path
+} // namespace pal

--- a/backends/qualcomm/runtime/pal/src/linux/DynamicLoading.cpp
+++ b/backends/qualcomm/runtime/pal/src/linux/DynamicLoading.cpp
@@ -1,0 +1,52 @@
+/*
+ * Copyright (c) Qualcomm Innovation Center, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <dlfcn.h>
+
+#include <pal/DynamicLoading.h>
+
+void* pal::dynamic_loading::DlOpen(const char* filename, int flags) {
+  int realFlags = 0;
+
+  if (flags & DL_NOW) {
+    realFlags |= RTLD_NOW;
+  }
+
+  if (flags & DL_LOCAL) {
+    realFlags |= RTLD_LOCAL;
+  }
+
+  if (flags & DL_GLOBAL) {
+    realFlags |= RTLD_GLOBAL;
+  }
+
+  if (flags & DL_NOLOAD) {
+#ifndef __hexagon__
+    realFlags |= RTLD_NOLOAD;
+#else
+    return nullptr;
+#endif
+  }
+
+  return ::dlopen(filename, realFlags);
+}
+
+void* pal::dynamic_loading::DlSym(void* handle, const char* symbol) {
+  return ::dlsym(handle, symbol);
+}
+
+int pal::dynamic_loading::DlClose(void* handle) {
+  if (!handle) {
+    return 0;
+  }
+  return ::dlclose(handle);
+}
+
+const char* pal::dynamic_loading::DlError() {
+  return ::dlerror();
+}

--- a/backends/qualcomm/runtime/pal/src/linux/Path.cpp
+++ b/backends/qualcomm/runtime/pal/src/linux/Path.cpp
@@ -1,0 +1,13 @@
+/*
+ * Copyright (c) Qualcomm Innovation Center, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <pal/Path.h>
+
+std::string pal::path::GetLibraryName(const std::string& baseName) {
+  return "lib" + baseName + ".so";
+}

--- a/backends/qualcomm/runtime/pal/src/windows/DynamicLoading.cpp
+++ b/backends/qualcomm/runtime/pal/src/windows/DynamicLoading.cpp
@@ -1,0 +1,86 @@
+/*
+ * Copyright (c) Qualcomm Innovation Center, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <windows.h>
+
+#include <cstring>
+
+#include <pal/DynamicLoading.h>
+
+static thread_local const char* sg_lastErrMsg = "";
+
+void* pal::dynamic_loading::DlOpen(const char* filename, int flags) {
+  if (!filename || std::strlen(filename) == 0) {
+    sg_lastErrMsg = "DlOpen: filename argument is null or empty";
+    return nullptr;
+  }
+
+  if (flags & DL_NOLOAD) {
+    HMODULE mod = GetModuleHandleA(filename);
+    if (!mod) {
+      sg_lastErrMsg = "DlOpen: library is not already loaded in the process";
+    }
+    return static_cast<void*>(mod);
+  }
+
+  if (!(flags & DL_NOW)) {
+    sg_lastErrMsg =
+        "DlOpen: flags must include DL_NOW for immediate resolution";
+    return nullptr;
+  }
+
+  HMODULE mod =
+      LoadLibraryExA(filename, nullptr, LOAD_WITH_ALTERED_SEARCH_PATH);
+  if (!mod) {
+    sg_lastErrMsg =
+        "DlOpen: LoadLibraryExA failed to load the specified library";
+    return nullptr;
+  }
+
+  return static_cast<void*>(mod);
+}
+
+void* pal::dynamic_loading::DlSym(void* handle, const char* symbol) {
+  FARPROC sym_addr = nullptr;
+
+  if ((!handle) || (!symbol)) {
+    sg_lastErrMsg = "DlSym: handle or symbol argument is null";
+    return nullptr;
+  }
+
+  HMODULE mod = static_cast<HMODULE>(handle);
+  sym_addr = GetProcAddress(mod, symbol);
+  if (!sym_addr) {
+    sg_lastErrMsg =
+        "DlSym: GetProcAddress failed to resolve the specified symbol";
+    return nullptr;
+  }
+
+  return reinterpret_cast<void*>(sym_addr);
+}
+
+int pal::dynamic_loading::DlClose(void* handle) {
+  if (!handle) {
+    return 0;
+  }
+
+  HMODULE mod = static_cast<HMODULE>(handle);
+
+  if (FreeLibrary(mod) == 0) {
+    sg_lastErrMsg = "DlClose: FreeLibrary failed to release the module handle";
+    return -1;
+  }
+
+  return 0;
+}
+
+const char* pal::dynamic_loading::DlError() {
+  const char* retStr = sg_lastErrMsg;
+  sg_lastErrMsg = "";
+  return retStr;
+}

--- a/backends/qualcomm/runtime/pal/src/windows/DynamicLoading.cpp
+++ b/backends/qualcomm/runtime/pal/src/windows/DynamicLoading.cpp
@@ -21,9 +21,15 @@ void* pal::dynamic_loading::DlOpen(const char* filename, int flags) {
   }
 
   if (flags & DL_NOLOAD) {
-    HMODULE mod = GetModuleHandleA(filename);
-    if (!mod) {
+    // Match the Linux RTLD_NOLOAD contract: return a handle for an
+    // already-loaded module with its reference count incremented, so the
+    // matching DlClose -> FreeLibrary stays balanced. Plain GetModuleHandleA
+    // would not take a reference, so FreeLibrary could prematurely unload a
+    // library still in use elsewhere.
+    HMODULE mod = nullptr;
+    if (!GetModuleHandleExA(0, filename, &mod)) {
       sg_lastErrMsg = "DlOpen: library is not already loaded in the process";
+      return nullptr;
     }
     return static_cast<void*>(mod);
   }

--- a/backends/qualcomm/runtime/pal/src/windows/Path.cpp
+++ b/backends/qualcomm/runtime/pal/src/windows/Path.cpp
@@ -1,0 +1,13 @@
+/*
+ * Copyright (c) Qualcomm Innovation Center, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <pal/Path.h>
+
+std::string pal::path::GetLibraryName(const std::string& baseName) {
+  return baseName + ".dll";
+}

--- a/backends/qualcomm/runtime/targets.bzl
+++ b/backends/qualcomm/runtime/targets.bzl
@@ -70,7 +70,7 @@ def define_common_targets():
                 ],
                 exclude = ["Logging.h"],
             ),
-            compiler_flags = [
+            exported_preprocessor_flags = [
                 "-Ibackends/qualcomm/runtime/pal/include",
             ],
             define_static_target = True,

--- a/backends/qualcomm/runtime/targets.bzl
+++ b/backends/qualcomm/runtime/targets.bzl
@@ -43,6 +43,8 @@ def define_common_targets():
             srcs = glob(
                 [
                     "*.cpp",
+                    "pal/src/linux/DynamicLoading.cpp",
+                    "pal/src/linux/Path.cpp",
                     "backends/*.cpp",
                     "backends/gpu/*.cpp",
                     "backends/htp/*.cpp",
@@ -58,6 +60,8 @@ def define_common_targets():
             exported_headers = glob(
                 [
                     "*.h",
+                    "pal/include/pal/DynamicLoading.h",
+                    "pal/include/pal/Path.h",
                     "backends/*.h",
                     "backends/gpu/*.h",
                     "backends/htp/*.h",
@@ -66,6 +70,9 @@ def define_common_targets():
                 ],
                 exclude = ["Logging.h"],
             ),
+            compiler_flags = [
+                "-Ibackends/qualcomm/runtime/pal/include",
+            ],
             define_static_target = True,
             link_whole = True,  # needed for executorch/examples/models/llama:main to register QnnBackend
             platforms = [ANDROID],

--- a/backends/qualcomm/utils/utils.py
+++ b/backends/qualcomm/utils/utils.py
@@ -5,6 +5,7 @@
 # LICENSE file in the root directory of this source tree.
 import operator
 import os
+import platform
 import re
 import warnings
 from collections import defaultdict, OrderedDict
@@ -75,6 +76,20 @@ from torch.export.exported_program import ExportedProgram
 from torch.fx import passes
 from torch.fx.passes.operator_support import OperatorSupportBase
 from torch.library import Library
+
+
+def _get_qnn_lib_name(base: str) -> str:
+    """Returns the platform-specific shared library filename for a QNN library."""
+    if platform.system().lower() == "windows":
+        return f"{base}.dll"
+    return f"lib{base}.so"
+
+
+def _get_qnn_host_lib_dir_name() -> str:
+    """Returns the QNN SDK library subdirectory name for the current x86-64 host OS."""
+    if platform.system().lower() == "windows":
+        return "x86_64-windows-msvc"
+    return "x86_64-linux-clang"
 
 
 class _AnnotationSkipper(OperatorSupportBase):
@@ -1208,7 +1223,7 @@ def generate_qnn_executorch_compiler_spec(  # noqa: C901
     qnn_executorch_options.dump_intermediate_outputs = dump_intermediate_outputs
 
     if saver:
-        qnn_executorch_options.library_path = "libQnnSaver.so"
+        qnn_executorch_options.library_path = _get_qnn_lib_name("QnnSaver")
         qnn_executorch_options.saver = True
         qnn_executorch_options.saver_output_dir = "saver_output"
 
@@ -1420,8 +1435,11 @@ def rewrite_prepared_observer(
 
 
 def get_sdk_build_id():
-    htp_library_path = (
-        os.environ.get("QNN_SDK_ROOT", None) + "/lib/x86_64-linux-clang/libQnnHtp.so"
+    htp_library_path = os.path.join(
+        os.environ.get("QNN_SDK_ROOT", None),
+        "lib",
+        _get_qnn_host_lib_dir_name(),
+        _get_qnn_lib_name("QnnHtp"),
     )
     # The GetQnnSdkBuildId API can be used without needing to create a backend first, so it works regardless of which backend is used.
     sdk_build_id = PyQnnManagerAdaptor.GetQnnSdkBuildId(htp_library_path)

